### PR TITLE
Abstract out emit_trap!

### PR DESCRIPTION
**Pull request by jpsamaroo**
Added v_readfirstlane_b32 workaround to GCN

### CODE REVIEW

**Comment by jpsamaroo** [comment]
Pairs with https://github.com/JuliaGPU/AMDGPUnative.jl/pull/52

**Comment by jpsamaroo** [comment]
@maleadt this is the fix to the weird traps/hangs I was always griping to you about. To summarize, GCN's `s_trap` instruction doesn't care about the active thread mask, and at least with LLVM 6, the AMDGPU target doesn't emit the right kind of branch to avoid the trap when the thread mask is zero (no active threads). This seems to have nothing to do with Julia's addrspaces "conflicting" with the target's addrspaces; sorry for constantly bugging you about that!

And big thanks to @vchuravy for pointing this out to me and showing me which parts of the AMDGPU target does the "right thing", and giving me hints on how to fix it!

File: src/gcn.jl [hunk 1]
```diff
@@ -136,3 +120,28 @@ function lower_throw_extra!(mod::LLVM.Module)
...
+    else
+        LLVM.Function(mod, "llvm.trap", LLVM.FunctionType(LLVM.VoidType(JuliaContext())))
+    end
+    rl_ft = LLVM.FunctionType(LLVM.Int32Type(JuliaContext()), [LLVM.Int32Type(JuliaContext())])
+    rl = if haskey(functions(mod), "llvm.amdgcn.readfirstlane")
+        functions(mod)["llvm.amdgcn.readfirstlane"]
+    else
+        LLVM.Function(mod, "llvm.amdgcn.readfirstlane", rl_ft)
+    end
+    # FIXME: Early versions of the AMDGPU target fail to skip machine blocks
```
Comment (vchuravy) [inline src/gcn.jl:136]: Check the version of LLVM and if you are above 9 just emit a trap?

**Comment by maleadt** [comment]
`lit` tests would be awesome to get rid of the regexes in `test`. Can we use `LLVM_full` for that?

**Comment by jpsamaroo** [comment]
@maleadt do you want me to wait on the lit tests to be ready to use, or can I add regexes for now to get this merged? I'd like to get JuliaGPU/AMDGPUnative.jl#52 merged soon to unblock other work (namely jpsamaroo/ROCArrays.jl#18).

**Comment by maleadt** [comment]
No, we won't get lit fixed anytime soon I think. Better merge this with some regexes for now.

**Pull request merged**